### ISSUE TITLE
[2.3] Probable bug while setting default and optional both

### ISSUE THREAD

**Issue by s7dhansh** [jagi/meteor-astronomy#534]
```
		listings: {
			type: ListingDetails,
			default: null,
			optional: true,
		},
```
This kind of declaration used to work (till 2.2.4). But in 2.3 it is causing a minimongo error on save:
```
MinimongoError: cannot use the part 'data' to traverse null [409]
```

The ListingDetails class contains an optional `data: [Object]` field. Since doc.listings is null, it is failing while finding `doc.listings.data` which ideally it should not.

~~Not sure if this is a Minimongo error or Astronomy.~~
The mnimongo was not updated between meteor 1.4.2 and 1.4.2.1, so I guess some new code in Astronomy caused this.
@jagi any ideas?

**Comment by s7dhansh** [comment]
```
each(Class.getFields(), (field) => {
    if (
      (field instanceof ObjectField) ||
      (field instanceof ListField && field.isClass)
    ) {
      doc[field.name] = field.castValue(doc[field.name], options);
    }
  });
```

In this code (castNested.js), you are casting each field to its class. So if I have a `listings` field as null, it too is being casted in its default Class. If `listings` is marked as optional, that should not happen right?

**Comment by lukejagodzinski** [comment]
What version of Astronomy are you using? There was 2.3.2 release which might fix this issue.

**Comment by s7dhansh** [comment]
2.3.2 only

**Comment by s7dhansh** [comment]
I am using 2.3.2 only

**Comment by lukejagodzinski** [comment]
Ok I will check it out, maybe there is still a bug in the cast method

**Comment by lukejagodzinski** [comment]
I've tested it and it works, so it's probably a bug in your code or you're missing some detail that may be crucial to know what's going on. The `castValue` method is not casting nil (null & undefined) values.

**Comment by lukejagodzinski** [comment]
Could you also post here a result of the `doc.getModifier()` call just before save?

**Comment by s7dhansh** [comment]
Check out this reprod repo: https://github.com/settlin/astro-bug

The relevant code:

```
import { Class } from "meteor/jagi:astronomy";

Posts = new Mongo.Collection("posts");

const Obj = Class.create({
    name: "Obj",
    fields: {
        a: {
            type: [Object],
            default() {return [];},
            optional: true
        },
    }
});

if (Meteor.isServer) {
    if (!Posts.findOne()) {
        Posts.insert({_id: "a"});
                **// inserting a null value for optional field**
        Posts.insert({_id: "b", obj: null});
    }
}

export const Post = Class.create({
    name: "Post",
    collection: Posts,
    secured: false,
    fields: {
        obj: {
            type: Obj,
            optional: true
        }
    }
});
```

On the client, on button click, I simply find it and save it (without modifying anything). Still I get modifiers.
console output:

```
Modifier for a:  {"$set":{"obj.a":[]}}
Modifier for b:  {"$set":{"obj.a":[]}}
Uncaught errorClass {error: 409, reason: "MinimongoError: cannot use the part 'a' to traverse null", details: undefined, message: "MinimongoError: cannot use the part 'a' to traverse null [409]",
```

Somehow, even if the data is `undefined` or `null`, there are modifiers which are causing an error.

**Comment by lukejagodzinski** [comment]
Thanks for reproduction. In deed there is a bug in the resolveValues method. I should resolve values for nested fields if a field is null or undefined. I'm preparing fix. Thanks for letting me know a bout this bug.

**Comment by lukejagodzinski** [comment]
Bug fixed in 2.3.3 release
